### PR TITLE
[Merged by Bors] - feat: SuccOrder and recursion principle on well-ordered type

### PR DESCRIPTION
**Pull request by alreadydone**
Order/SuccPred/Basic.lean: add `SuccOrder.ofLinearWellFoundedLT`, which puts a SuccOrder on an arbitrary well-ordered type. Also add the dual, `PredOrder` version. Add missing `pred` lemmas corresponding to existing `succ` lemmas.

Order/SuccPred/Limit.lean: add `SuccOrder.limitRecOn` generalizing `Ordinal.limitRecOn` to allow definition by transfinite recursion on an arbitrary well-ordered type. (It turns out a **partial** well-founded order is enough.)

SetTheory/Ordinal/Arithmetic.lean: refactor `Ordinal.limitRecOn` to use `SuccOrder.limitRecOn`.

SetTheory/Cardinal/Ordinal.lean: add a lemma saying that the initial ordinal of an infinite cardinal is a NoMaxOrder.

SetTheory/Ordinal/Basic.lean: add `mk_Iio_ord_out_α`, a restatement of `card_typein_out_lt`.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by alreadydone** [comment]
cc @digama0, the [author](https://github.com/leanprover-community/mathlib/commit/7726a92f) of `ordinal.limit_rec_on`.

File: Mathlib/SetTheory/Ordinal/Arithmetic.lean [hunk 1]
```diff
@@ -240,10 +240,14 @@ def IsLimit (o : Ordinal) : Prop :=
   o ≠ 0 ∧ ∀ a < o, succ a < o
 #align ordinal.is_limit Ordinal.IsLimit
 
+theorem IsLimit.isSuccLimit {o} (h : IsLimit o) : IsSuccLimit o := isSuccLimit_iff_succ_lt.mpr h.2
```
Comment (alreadydone) [inline Mathlib/SetTheory/Ordinal/Arithmetic.lean:243]: There is an asymmetry between the definitions
```
Ordinal.IsLimit o = (o ≠ 0 ∧ ∀ a < o, Order.succ a < o)
Cardinal.IsLimit c = (c ≠ 0 ∧ Order.IsSuccLimit c)
```
but it's probably out of scope of this PR to align them.

**Review by YaelDillies — COMMENTED** [review]
The rest looks good

**Comment by jcommelin** [comment]
bors d+

**Comment by alreadydone** [comment]
Thanks for the review and delegation!

File: Mathlib/Order/SuccPred/Limit.lean [hunk 2]
```diff
@@ -185,6 +185,40 @@ theorem isSuccLimitRecOn_succ' (hs : ∀ a, ¬IsMax a → C (succ a)) (hl : ∀
...
+  have h := not_isSuccLimit_succ_of_not_isMax ha
+  rw [SuccOrder.limitRecOn, WellFounded.fix_eq, dif_neg h]
+  have {b c hb hc} {x : ∀ a, C a} (h : b = c) :
+    congr_arg succ h ▸ H_succ b hb (x b) = H_succ c hc (x c) := by subst h; rfl
+  let x := Classical.indefiniteDescription _ (not_isSuccLimit_iff.mp h)
+  exact this ((succ_eq_succ_iff_of_not_isMax x.2.1 ha).mp x.2.2)
+
+@[simp]
+theorem _root_.SuccOrder.limitRecOn_limit (ha : IsSuccLimit a) : SuccOrder.limitRecOn a H_succ H_lim
+    = H_lim a ha fun x _ ↦ SuccOrder.limitRecOn x H_succ H_lim := by
```
[the remark this answers is not in the document]
Comment (alreadydone) [reply Mathlib/Order/SuccPred/Limit.lean]: Forgot to say: omitting the namespace doesn't actually work, which may have to do with the `_root_`.

[the remark this answers is not in the document]
Comment (YaelDillies) [reply Mathlib/Order/SuccPred/Limit.lean]: Oh yeah that's a known bug

**Pull request closed without merging**